src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import linear_algebra.matrix.orthogonal
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext i j,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h] }
+end
```

Comment:
It's shorter if you prove `is_diag.diagonal` first:
```suggestion
@[simp] lemma is_diag.diagonal [has_zero α] [decidable_eq n] (d : n → α) :
  (diagonal d).is_diag :=
λ i j, matrix.diagonal_apply_ne

/-- Diagonal matrices are generated by `matrix.diagonal`. -/
lemma is_diag.exists_diagonal [has_zero α] [decidable_eq n] {A : matrix n n α} (h : A.is_diag) :
  ∃ d, diagonal d = A :=
begin
  refine ⟨λ i, A i i, ext $ λ i j, _⟩,
  obtain rfl | hij := decidable.eq_or_ne i j,
  { rw diagonal_apply_eq },
  { rw [diagonal_apply_ne hij, h hij] },
end

/-- `matrix.is_diag.exists_diagonal` as an iff. -/
lemma is_diag_iff_exists_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
  A.is_diag ↔ (∃ d, diagonal d = A) :=
⟨is_diag.exists_diagonal, λ ⟨d, hd⟩, hd ▸ is_diag.diagonal d⟩
```
I've made two other changes above:
* Added `is_diag.exists_diagonal` for dot notation
* Swapped the direction of the `=` so that `obtain ⟨d, rfl⟩ := h.exists_diagonal` works.